File: HTProject_OC/CollectionView/CollectionViewController.m
//
//  CollectionViewController.m
//  HTProject_OC
//
//  Created by Hem1ng on 2019/12/9.
//  Copyright © 2019 Hem1ngT4i. All rights reserved.
//

#import "CollectionViewController.h"
#import "HTCollectionViewFlowLayout.h"

@interface CollectionViewController ()<UICollectionViewDelegate,UICollectionViewDataSource>

@property (strong, nonatomic) UICollectionView *collectionView;

@end

@implementation CollectionViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    HTCollectionViewFlowLayout *layout = [[HTCollectionViewFlowLayout alloc] init];
    CGFloat fSpacing = 10;
    layout.minimumInteritemSpacing = fSpacing;
    layout.minimumLineSpacing = fSpacing;
    layout.sectionInset = UIEdgeInsetsMake(fSpacing, fSpacing, fSpacing, fSpacing);
    //创建变量数组
    NSMutableArray *delatArray = [NSMutableArray array];
    for (NSInteger i = 0; i < 50; i++) {
        [delatArray addObject:[NSNumber numberWithDouble:40 + arc4random() % 40]];
    }
    [layout flowLayoutWithItemSize:CGSizeMake(60, 60) deltaArray:delatArray style:HTAlignmentLayoutLeft];
    self.collectionView = [[UICollectionView alloc] initWithFrame:CGRectMake(0, 0, kScreenWidth, kScreenHeight - 83 - 88) collectionViewLayout:layout];
    self.collectionView.backgroundColor = UIColor.blackColor;
    self.collectionView.delegate = self;
    self.collectionView.dataSource = self;
    [self.collectionView registerClass:[UICollectionViewCell class] forCellWithReuseIdentifier:@"HTCell"];
    [self.view addSubview:self.collectionView];
}

- (NSInteger)collectionView:(UICollectionView *)collectionView numberOfItemsInSection:(NSInteger)section {
    return 50;
}

- (UICollectionViewCell *)collectionView:(UICollectionView *)collectionView cellForItemAtIndexPath:(NSIndexPath *)indexPath {
    UICollectionViewCell *cell = [collectionView dequeueReusableCellWithReuseIdentifier:@"HTCell" forIndexPath:indexPath];
    UILabel *label = (UILabel *)[cell viewWithTag:10];
    if (!label) {
        [cell setBackgroundColor:[UIColor orangeColor]];
        label = [[UILabel alloc]init];
        label.textAlignment = NSTextAlignmentCenter;
        label.textColor = UIColor.whiteColor;
        label.font = [UIFont systemFontOfSize:30];
        label.adjustsFontSizeToFitWidth = YES;
        label.tag = 10;
        [cell.contentView addSubview:label];
    }
    [label setFrame:CGRectMake(0, 0, cell.frame.size.width, cell.frame.size.height)];
    label.text = [[NSString alloc] initWithFormat:@"%ld", indexPath.item + 1];
    return cell;
}

- (void)collectionView:(UICollectionView *)collectionView didSelectItemAtIndexPath:(NSIndexPath *)indexPath {
    NSLog(@"--%@",indexPath);
}
@end

File: HTProject_Swift/Demos/CollectionViewController.swift
//
//  CollectionViewController.swift
//  HTProject_Swift
//
//  Created by Hem1ngT4i on 2024/1/11.
//  Copyright © 2024 Hem1ngT4i. All rights reserved.
//

import UIKit

class CollectionViewController: UIViewController, UICollectionViewDataSource, UICollectionViewDelegate {
    private let reuseId = "ReusedId"
    let dataList = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16]
    
    override func viewDidLoad() {
        super.viewDidLoad()
        view.backgroundColor = .white
        
        let layout = UICollectionViewFlowLayout()
        let screenWidth = UIScreen.main.bounds.size.width
        let itemSpacing: CGFloat = 50
        let width = (screenWidth - itemSpacing) / 2.0
        layout.itemSize = CGSize(width: width, height: 120)
        layout.minimumLineSpacing = 10
        layout.minimumInteritemSpacing = itemSpacing
        layout.scrollDirection = .horizontal
        let collectionView = UICollectionView(frame: CGRect(x: 0, y: 0, width: screenWidth, height: 140), collectionViewLayout: layout)
        collectionView.center = view.center
        collectionView.backgroundColor = .green
        collectionView.dataSource = self
        collectionView.delegate = self
        collectionView.register(UICollectionViewCell.self, forCellWithReuseIdentifier: reuseId)
        view.addSubview(collectionView)
    }
    
    func numberOfSections(in collectionView: UICollectionView) -> Int {
        1
    }

    func collectionView(_ collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
        dataList.count
    }
    
    func collectionView(_ collectionView: UICollectionView, cellForItemAt indexPath: IndexPath) -> UICollectionViewCell {
        let cell = collectionView.dequeueReusableCell(withReuseIdentifier: reuseId, for: indexPath)
        let view1001 = cell.contentView.subviews.first { $0.tag == 1001 }
        if view1001 == nil {
            let label = UILabel(frame: CGRect(x: 0, y: 0, width: 100, height: 100))
            label.backgroundColor = .red
            label.tag = 1001
//            label.image = UIImage(systemName: "circle")
            label.text = "\(dataList[indexPath.row])"
            cell.contentView.addSubview(label)
        }
        cell.backgroundColor = .orange
        return cell
    }
    
    func collectionView(_ collectionView: UICollectionView, didSelectItemAt indexPath: IndexPath) {
        print("indexPath: \(indexPath)")
    }
}
